app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt

```diff
@@ -0,0 +1,119 @@
+
+    @JvmStatic
+    fun get(wiki: WikiSite): Service {
+        val hashCode = wiki.hashCode().toLong()
+        if (SERVICE_CACHE[hashCode] != null) {
+            return SERVICE_CACHE[hashCode]!!
+        }
+        val r = createRetrofit(wiki, getBasePath(wiki))
+        val s = r.create(Service::class.java)
+        SERVICE_CACHE.put(hashCode, s)
+        return s
+    }
+
+    fun getRest(wiki: WikiSite): RestService {
+        val hashCode = wiki.hashCode().toLong()
+        if (REST_SERVICE_CACHE[hashCode] != null) {
+            return REST_SERVICE_CACHE[hashCode]!!
+        }
+        val r = createRetrofit(wiki, getRestBasePath(wiki))
+        val s = r.create(RestService::class.java)
+        REST_SERVICE_CACHE.put(hashCode, s)
+        return s
+    }
+
+    fun getCoreRest(wiki: WikiSite): CoreRestService {
+        val hashCode = wiki.hashCode().toLong()
+        if (CORE_REST_SERVICE_CACHE[hashCode] != null) {
+            return CORE_REST_SERVICE_CACHE[hashCode]!!
+        }
+        val r = createRetrofit(wiki, wiki.url() + "/" + CoreRestService.CORE_REST_API_PREFIX)
+        val s = r.create(CoreRestService::class.java)
+        CORE_REST_SERVICE_CACHE.put(hashCode, s)
+        return s
+    }
+
+    @JvmStatic
+    fun getAnalyticsRest(streamConfig: StreamConfig): EventService {
+        val destinationEventService = streamConfig.destinationEventService
+        if (ANALYTICS_REST_SERVICE_CACHE[destinationEventService.id] != null) {
+            return ANALYTICS_REST_SERVICE_CACHE[destinationEventService.id]!!
+        }
+        val intakeBaseUriOverride = Prefs.getEventPlatformIntakeUriOverride().orEmpty().ifEmpty { destinationEventService.baseUri }
+        val r = createRetrofit(null, intakeBaseUriOverride)
+        val s = r.create(EventService::class.java)
+        ANALYTICS_REST_SERVICE_CACHE.put(destinationEventService.id, s)
+        return s
+    }
+
+    operator fun <T> get(wiki: WikiSite, baseUrl: String?, service: Class<T>?): T {
+        val r = createRetrofit(wiki, (if (baseUrl.isNullOrEmpty()) wiki.url() + "/" else baseUrl))
+        return r.create(service)
+    }
+
+    private fun getBasePath(wiki: WikiSite): String {
+        return if (Prefs.getMediaWikiBaseUrl().isEmpty()) wiki.url() + "/" else Prefs.getMediaWikiBaseUrl()
+    }
+
+    fun getRestBasePath(wiki: WikiSite): String {
+        var path = if (Prefs.getRestbaseUriFormat().isEmpty()) wiki.url() + "/" + RestService.REST_API_PREFIX
+        else String.format(Prefs.getRestbaseUriFormat(), "https", wiki.authority())
+        if (!path.endsWith("/")) {
+            path += "/"
+        }
+        return path
+    }
+
+    private fun createRetrofit(wiki: WikiSite?, baseUrl: String): Retrofit {
+        val okHttpClientBuilder = client.newBuilder()
+        if (wiki != null) {
+            okHttpClientBuilder.addInterceptor(LanguageVariantHeaderInterceptor(wiki))
+        }
```

Comment:
Very curious that why we have to add the `LanguageVariantHeaderInterceptor` here, and then add it again below? Do you have any reason why we are doing this?